project_euler/problem_059/sol1.py

```diff
@@ -0,0 +1,128 @@
+be memorable.
+
+Your task has been made easy, as the encryption key consists of three lower case
+characters. Using p059_cipher.txt (right click and 'Save Link/Target As...'), a
+file containing the encrypted ASCII codes, and the knowledge that the plain text
+must contain common English words, decrypt the message and find the sum of the ASCII
+values in the original text.
+"""
+
+
+import string
+from itertools import cycle, product
+from pathlib import Path
+from typing import List, Optional, Set, Tuple
+
+VALID_CHARS: str = (
+    string.ascii_letters + string.digits + string.punctuation + string.whitespace
+)
+LOWERCASE_INTS: List[int] = [ord(letter) for letter in string.ascii_lowercase]
+VALID_INTS: Set[int] = {ord(char) for char in VALID_CHARS}
+
+COMMON_WORDS: List[str] = ["the", "be", "to", "of", "and", "in", "that", "have"]
+
+
+def try_key(ciphertext: List[int], key: Tuple[int, ...]) -> Optional[str]:
+    """
+    Given an encrypted message and a possible 3-character key, decrypt the message.
+    If the decrypted message contains a invalid character, i.e. not an ASCII letter,
+    a digit, punctuation or whitespace, then we know the key is incorrect, so return
+    None.
+    >>> try_key([0, 17, 20, 4, 27], (104, 116, 120))
+    'hello'
+    >>> try_key([68, 10, 300, 4, 27], (104, 116, 120)) is None
+    True
+    """
+    decoded: str = ""
+    keychar: int
+    cipherchar: int
+    decodedchar: int
+
+    for keychar, cipherchar in zip(cycle(key), ciphertext):
+        decodedchar = cipherchar ^ keychar
+        if decodedchar not in VALID_INTS:
+            return None
+        decoded += chr(decodedchar)
+
+    return decoded
+
+
+def filter_valid_chars(ciphertext: List[int]) -> List[str]:
+    """
+    Given an encrypted message, test all 3-character strings to try and find the
+    key. Return a list of the possible decrypted messages.
+    >>> from itertools import cycle
+    >>> text = "The enemy's gate is down"
+    >>> key = "end"
+    >>> encoded = [ord(k) ^ ord(c) for k,c in zip(cycle(key), text)]
+    >>> text in filter_valid_chars(encoded)
+    True
+    """
+    possibles: List[str] = []
+    for key in product(LOWERCASE_INTS, repeat=3):
+        encoded = try_key(ciphertext, key)
+        if encoded is not None:
+            possibles.append(encoded)
+    return possibles
+
+
+def filter_common_word(possibles: List[str], common_word: str) -> List[str]:
+    """
+    Given a list of possible decoded messages, narrow down the possibilities
+    for checking for the presence of a specified common word. Only decoded messages
+    containing common_word will be returned.
+    >>> filter_common_word(['asfla adf', 'I am here', '   !?! #a'], 'am')
+    ['I am here']
+    >>> filter_common_word(['athla amf', 'I am here', '   !?! #a'], 'am')
+    ['athla amf', 'I am here']
+    """
+    return [possible for possible in possibles if common_word in possible.lower()]
+
+
+def solution() -> int:
```

Comment:
@dhruvmanila I can't see any way to provide a doctest for this function since the input is read from a file. Advice?

Reply:
If possible, add another text file containing test input and write a doctest for that file. A parameter would be required for the function whose default value will be the original text file.

This is not compulsory so if you plan not to do it inform me so that I can merge this.